src/firestore/indexes.ts

```diff
@@ -0,0 +1,560 @@
+      return false;
+    }
+
+    if (index.fields.length !== spec.fields.length) {
+      return false;
+    }
+
+    let i = 0;
+    while (i < index.fields.length) {
+      const iField = index.fields[i];
+      const sField = spec.fields[i];
+
+      if (iField.fieldPath !== sField.fieldPath) {
+        return false;
+      }
+
+      if (iField.order !== sField.order) {
+        return false;
+      }
+
+      if (iField.arrayConfig !== sField.arrayConfig) {
+        return false;
+      }
+
+      i++;
+    }
+
+    return true;
+  }
+
+  /**
+   * Determine if an API Field and a Spec Field are functionally equivalent.
+   */
+  fieldMatchesSpec(field: API.Field, spec: Spec.FieldOverride): boolean {
+    const parsedName = this.parseFieldName(field.name);
+
+    if (parsedName.collectionGroupId !== spec.collectionGroup) {
+      return false;
+    }
+
+    if (parsedName.fieldPath !== spec.fieldPath) {
+      return false;
+    }
+
+    if (field.indexConfig.indexes.length !== spec.indexes.length) {
+      return false;
+    }
+
+    const fieldModes = field.indexConfig.indexes.map((index) => {
+      const firstField = index.fields[0];
+      return firstField.order || firstField.arrayConfig;
+    });
+
+    const specModes = spec.indexes.map((index) => {
+      return index.order || index.arrayConfig;
+    });
+
+    // Confirms that the two objects have the same set of enabled indexes without
+    // caring about specification order.
+    for (const mode of fieldModes) {
+      if (specModes.indexOf(mode) < 0) {
+        return false;
+      }
+    }
+
+    return true;
+  }
+
+  /**
+   * Parse an Index name into useful pieces.
+   */
+  parseIndexName(name: string | undefined): IndexName {
+    if (!name) {
+      throw new FirebaseError(`Cannot parse undefined index name.`);
+    }
+
+    const m = name.match(INDEX_NAME_REGEX);
+    if (!m || m.length < 4) {
+      throw new FirebaseError(`Error parsing index name: ${name}`);
+    }
+
+    return {
+      projectId: m[1],
+      collectionGroupId: m[2],
+      indexId: m[3],
+    };
+  }
+
+  /**
+   * Parse an Field name into useful pieces.
+   */
+  parseFieldName(name: string): FieldName {
+    if (!name) {
+      throw new FirebaseError(`Cannot parse undefined field name.`);
+    }
+
+    const m = name.match(FIELD_NAME_REGEX);
+    if (!m || m.length < 4) {
+      throw new FirebaseError(`Error parsing field name: ${name}`);
+    }
+
+    return {
+      projectId: m[1],
+      collectionGroupId: m[2],
+      fieldPath: m[3],
+    };
+  }
+
+  /**
+   * Take a object that may represent an old v1beta1 indexes spec
+   * and convert it to the new v1beta2/v1 spec format.
+   *
+   * This function is meant to be run **before** validation and
+   * works on a purely best-effort basis.
+   */
+  upgradeOldSpec(spec: any): any {
+    const result = {
+      indexes: [],
+      fieldOverrides: spec.fieldOverrides || [],
+    };
+
+    if (!spec.indexes) {
+      return;
```

Comment:
I feel like this is a subtle bug waiting to happen. Can this throw an error, or return an empty object, log something...
?

Reply:
This method is explicitly designed to not fail hard so it can be called before validation and then the validator can assume it's validating based on v1beta2 rules.  The first step in `validateSpec` is to throw if there's no `indexes` field.